frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -752,94 +736,82 @@ export default ({
+          sinceHeight = Math.max(0, this.messages[0].height - limit)
         }
+        events = await collectEventStream(sbp('chelonia/out/eventsAfter', chatRoomID, sinceHeight, latestHeight - sinceHeight + 1))
       }
 
-      if (!this.checkEventSourceConsistency(chatRoomId)) return
+      if (!this.checkEventSourceConsistency(chatRoomID)) return
 
-      if (!isLoadedFromStorage) {
-        // NOTE: already rendered above in this function
+      if (events.length) {
         await this.rerenderEvents(events)
-        if (!this.checkEventSourceConsistency(chatRoomId)) return
       }
 
       if (!this.ephemeral.messagesInitiated) {
         this.setStartNewMessageIndex()
-        this.updateScroll(messageHashToScroll, Boolean(mhash)) // We do want the 'c-focused' animation if there is a message-scroll query.
-        return false
+
+        // NOTE: we do want the 'c-focused' animation if there is a message-scroll query.
+        if (events.length) {
+          // NOTE: if 'messageHashToScroll' was not there in the messages of the contract state
+          //       we need to retrieve more events, and render to scroll to that message
+          this.updateScroll(messageHashToScroll, Boolean(mhash))
+        } else {
+          // NOTE: we need to scroll to the message first in order to no more infiniteHandler is called
+          await this.updateScroll(messageHashToScroll, Boolean(mhash))
+        }
       }
 
-      return events.length < limit
+      return events.length > 0 && GIMessage.deserializeHEAD(events[0]).head.height === 0
     },
     rerenderEvents (events) {
-      const contractID = this.summary.chatRoomId
-
-      if (!this.ephemeral.messagesInitiated || this.latestEvents.length === 0) {
+      if (!this.latestEvents.length) {
         this.latestEvents = events
       } else if (events.length > 1) {
-        events.pop() // remove duplication. For more detail, check sbp('chelonia/out/eventsBetween')
         this.latestEvents.unshift(...events)
       }
 
+      const contractID = this.summary.chatRoomID
       // This ensures that `this.latestEvents.push(event)` below happens in order
       return sbp('okTurtles.eventQueue/queueEvent', CHATROOM_EVENTS, async () => {
         if (!this.checkEventSourceConsistency(contractID)) return
 
         if (this.latestEvents.length > 0) {
-          let state = this.generateNewChatRoomState()
+          let state = this.generateNewChatRoomState(true)
+          const entryHeight = GIMessage.deserializeHEAD(this.latestEvents[0]).head.height
+          if (entryHeight > 0) {
+            // NOTE: considering the KeyAdd GI event
+            //       authorizedKeys should not exist before _notBeforeHeight
+            Object.keys(state._vm.authorizedKeys).forEach(keyId => {
+              if (state._vm.authorizedKeys[keyId]._notBeforeHeight >= entryHeight) {
+                delete state._vm.authorizedKeys[keyId]
+              }
+            })
+          }
+
```

Comment:
@corrideat could you review these and related changes to `ChatMain.vue` and give either your approval or a change request?

Reply:
@Silver-IT Can you explain your reasoning here? I don't think you should be doing this at all, for three reasons:

1. The `_notBeforeHeight` and `_notAfterHeight` attributes are checked (or, at least, should be; if they aren't, it's a bug) by Chelonia everywhere where they matter. This means that having 'extra' keys should not have any effect.
2. The goal is for contract developers not to have to mess with Chelonia internals, for the most part. Doing manual key management like this is at odds with that, and it also makes it difficult to maintain (for example, in the future we might change how the `not*Height` thing works to support adding and removing keys with different permissions throughout their lifetime) if there are future changes in Chelonia.
3. The work that I'm currently doing (moving Chelonia into a service worker) will more strictly separate Chelonia from the UI. While it's not clear yet that this particular addition won't work with the new architecture, it certainly goes against the principle of keeping things separate.

Besides the reasons above, I'm not sure why you're doing this. If you could explain, I could help see if there's another and better way to go about it.

Reply:
Other than that, there's an off-by-one error. If you want to remove keys in the future, the check should be `_notBeforeHeight > entryHeight`. As the name says, _not before_, the range is inclusive. So, a key with `_notBeforeHeight` of `x` is valid for height `x`, but _not before_ that. So, it's not valid for heights `[0, x-1]`.

Reply:
I think I see it now. Are you trying to remove keys from the future so that key operations for those keys do get processed? If so, that makes some sense, but doesn't it still work without it? ]

Reply:
> I think I see it now. Are you trying to remove keys from the future so that key operations for those keys do get processed?

Yeah, since those keys would be added again when the future events are processed. I described this in the comment, but if it doesn't make sense, we should update it.

> If so, that makes some sense, but doesn't it still work without it?

I remember it made errors (I don't remember the error name now), and that's why I made additional work to remove those keys.

Reply:
@Silver-IT could you please share more info about what errors you encountered?

I very much agree with @corrideat that this code shouldn't be here because we don't want developers to have to mess around with Chelonia internals in order to implement their app.

So if you can share the steps plus any errors that you encounter, it would help us figure out how to properly address this.

Reply:
> if you can share the steps plus any errors that you encounter, it would help us figure out how to properly address this.

Steps to reproduce
- [Tab1] user1 creates group, and create a new channel `channel-1`
- [Tab1] user1 sends more than 60 messages. (when I was testing, I sent 70 messages)
- [Tab2] user2 joins the group, and joins the channel `channel-1`
- [Tab2] user2 switches to general chatroom, and switches back to `channel-`
- [Tab2] user2 scrolls up so that infinite scroll would work
- [Tab2] user2 would get the error in the console
![image](https://github.com/okTurtles/group-income/assets/57976479/4c84930d-e184-44a8-8a8c-dd86486bb6e7)

This error happens inside the `keysToMap` function in `chelonia/internals.js`.

Reply:
How was this not an issue in the past? In either case, this is something that could be fixed in Chelonia.

Option A) Create a selector that does pretty much what you're doing here (actually, I've already defined a new selector (not in master) to get the contract state for a contract, it could also take a height parameter and remove keys.
Option B) Change the validation code that's failing to take _notBefore into account. If `_notBefore > thisMsgHeight`, then treat the key as not existing.

Reply:
@Silver-IT Take a look at https://github.com/okTurtles/group-income/pull/1993, implementing option A.

You can incorporate those changes into this PR using the web UI or in the terminal by using `git merge` or `git cherry-pick`.